### DISCUSSION TITLE
Question about joint constraints and physics bodies.

### DISCUSSION THREAD

**Discussion by venumspyder** [jrouwe/JoltPhysics#933] [Q&A] [answered]
So I have a rag doll set up in the game. I have set up the rag doll using the RagdollSettings and defined the WORLD SPACE positions and orientations of the bodies and joint constraints.

I've noticed that when you call setPose in the rag doll class it loops through all the rag doll bodies and sets the orientation and position for each of the bodies. However, the pose I am passing in consists of the joint positions and orientations for a specific frame of animation. How does Jolt place the bodies in the correct location and orientation if all I am passing in is the pose for the skeleton joints of an animation frame? I assume it accounts for this with the information I have passed in with RagdollSettings when creating the rag doll?

And a similar question is when I call GetPose in the rag doll class, it loops through the bodies that are registered with the rag doll and returns the position and orientation of each body minus the root offset. While this is good if you are placing geometric representations of the bodies that make up the rag doll, I actually need the joint positions and orientations of the skeleton. I plan to use that to then skin a character mesh in the game. Is there a way I can get the joint position\orientation for the rag doll skeleton as opposed to the bodies?

**Comment by jrouwe** [comment]
> How does Jolt place the bodies in the correct location and orientation if all I am passing in is the pose for the skeleton joints of an animation frame?

The assumption is that the body position/orientation IS the joint position/orientation. E.g. a capsule shape would be offset using a RotatedTranslatedShape so that it is correctly placed if you put the body on the joint.

**Comment by venumspyder** [comment]
OK assume I have a leg joint as **LThigh** and it is positioned in world space at 

constraint_position **0.0934882 0.332284 0.0**		// LThigh

I then have a capsule shape positioned in world space at:

shape_position **0.1 0.21 -0.04**				// LThigh

Here is a picture:

![Jolt](https://github.com/jrouwe/JoltPhysics/assets/139675035/993d6df9-5924-4320-bd5d-6260c3e46d24)

So currently I am assuming that this is NOT possible since the capsule shape is not centered around the **LThigh** joint as you
suggest and based on what I have seen in the source code.

Currently I set the rag doll body like so:

       RagdollSettings::Part &rcPart = cRagdollSettings->mParts [iPart];

       // set the world space position of the body here.....
       rcPart.mPosition.Set (0.1f, 0.21f, -0.04f);

        if (iPart > 0)
        {
            SwingTwistConstraintSettings *pcConstraint = new SwingTwistConstraintSettings ();
            vsAssert (pcConstraint != NULL);
            
            pcConstraint->mDrawConstraintSize = 0.1f;
            
            // set the skeleton joint constraint
            pcConstraint->mPosition1.Set (0.0934882f, 0.332284f, 0.0f);
            
            pcConstraint->mPosition2.Set (0.0934882f, 0.332284f, 0.0f);
        }

Based on your statements, this is an incorrect set up that will not get me the desired result. I also noticed that in your **CreateRigTest.cpp** example your capsule shapes **positions[]** are offset from the **constraint_positions[]** defined in the file. Of course if you are only drawing the body shapes (as in the example) all is well. However, this won't work well with an animated polygonal mesh.

Is there no way around this? ie. I HAVE to have the capsule shape in the exact same position as the joint even if it doesn't match the polygonal mesh to get this to work?

I don't see how setting it like this:

```
capsule_shape = new CapsuleShape ();

RotatedTranslatedShapeSettings(Vec3 (0.1f, 0.21f, -0.04f), QuatArg inRotation, capsule_shape);
```

And then setting the part position as well as the swing twist constraint settings to be the same position of **0.0934882 0.332284 0.0** would make any difference in what I am trying to achieve. For example:

```
RagdollSettings::Part &rcPart = cRagdollSettings->mParts [iPart];

   // set the world space position of the body here SAME AS THE SwingTwistConstraintSettings.....
   rcPart.mPosition.Set (0.0934882f, 0.332284f, 0.0f);

    if (iPart > 0)
    {
        SwingTwistConstraintSettings *pcConstraint = new SwingTwistConstraintSettings ();
        vsAssert (pcConstraint != NULL);
        
        pcConstraint->mDrawConstraintSize = 0.1f;
        
        // set the skeleton joint constraint
        pcConstraint->mPosition1.Set (0.0934882f, 0.332284f, 0.0f);
        
        pcConstraint->mPosition2.Set (0.0934882f, 0.332284f, 0.0f);
    }
```

Perhaps this is the correct way to set this up for my use case. If it is not,  could you please provide some really simple pseudo code to set up my rag doll LThigh joint example? Thanks in advance for any help you can provide!

**Comment by jrouwe** [comment, marked as answer]
Instead of:

```
RotatedTranslatedShapeSettings(Vec3 (0.1f, 0.21f, -0.04f), QuatArg inRotation, capsule_shape);
```

you need:

```
RotatedTranslatedShapeSettings(Vec3(0.1f, 0.21f, -0.04f) - Vec3(0.0934882f, 0.332284f, 0.0f), Quat::sIdentity(), capsule_shape);
```

You want to translate the shape so that if you place the body on the joint position that the capsule is in the right position.

**Reply by venumspyder** [reply]
Thanks for the quick reply! Now I understand it!

**Comment by venumspyder** [comment]
One thing I forgot to mention. For the arms since they have been rotated and are capsules the code would be (Assuming same positions as the legs described above):

```
RotatedTranslatedShapeSettings(Vec3(0.1f, 0.21f, -0.04f) - Vec3(0.0934882f, 0.332284f, 0.0f), 
                                                      Quat::sRotation(Vec3::sAxisZ(), 0.5f * JPH_PI), 
                                                      capsule_shape);
```

Or do I set the rotation in the part as shown in **CreateRigTest.cpp** instead?

```
rotations[p] = Quat::sRotation(Vec3::sAxisZ(), 0.5f * JPH_PI);

		RagdollSettings::Part &part = settings->mParts[p];
		part.SetShape(shapes[p]);
		part.mPosition = positions[p];
		part.mRotation = rotations[p];
```

**Reply by jrouwe** [reply]
Same as with the positions, you need to use the relative orientation between the joint and the capsule orientation:

```
Quat joint_orientation = ...;
Quat capsule_orientation = ...;
Quat relative_orientation = joint_orientation.GetConjugated() * capsule_orientation;
```

but note that in that case you also need to make the translation relative to the joint orientation:

```
Quat joint_translation = ...;
Quat capsule_translation = ...;
Quat relative_orientation = joint_orientation.GetConjugated() * (capsule_translation - joint_translation);
```

and then you can use:

```
RotatedTranslatedShapeSettings(relative_translation, relative_orientation, capsule_shape);
```
